Splits off the logic from sharees endpoint thus making it available from within Nc/via PHP.

blizzz opened the pull request:
@schiessle as discussed yesterday.  Took a bit more effort, because i decided to split off the different getters to Plugins. However, I did not create a register function or anything. Can be added later if it becomes necessary. Right now, I doubt this and want to keep effort small.  I did not touch tests, yet, however Smoke tests with the share dialog worked for me. Are you OK with that approach?

necessary for #2443 

P.S.: Context: @schiessle prefered to have the logic outside of the Sharing scope, due to his concern that a universal API is established again. Therefore we thought the Collaborator terminology could fit.

blizzz committed:
Splits off the logic from sharees endpoint thus making it available from
within Nc/via PHP.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

daita commented:
could be fun to merge with https://github.com/nextcloud/server/pull/5280

blizzz commented:
on first glance, i did not see conflict potential. different controller and endpoint.

blizzz committed:
update autoloader

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
missing PHP doc for public interfaces

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

nickvergessen commented on lib/private/Collaboration/Collaborators/CirclePlugin.php:
```diff
@@ -0,0 +1,49 @@
+use OCP\Collaboration\Collaborators\ISearchPlugin;
+use OCP\Collaboration\Collaborators\ISearchResult;
+
+class CirclePlugin implements ISearchPlugin {
```
should be provided by the circles app?

blizzz replied:
Yes, then I need to add a register method and such… or another entry for the appinfo xml file… not much of a big deal, of course, I was hoping to get around this. After all, it's "just" axe sharpening for the autocomplete.

blizzz replied:
The annoying part here is the shareType. The client needs to request the shareTypes as they are define in \OC\Share\Constants (made available via \OCP\Share). This makes it hard to register new share types for apps, and even harder for clients for figuring out new features. OTOH all the types may have special aspects themselves, so special treatment might be necessary nevertheless. 

That said, what makes most sense to me is to allow registering plugins for existing share types. And opening this for 1:x relationships (one type, many plugins).

nickvergessen commented on lib/private/Collaboration/Collaborators/MailPlugin.php:
```diff
@@ -0,0 +1,162 @@
+			$result['wide'] = [];
+		}
+
+		if (!$searchResult->hasExactIdMatch('emails') && filter_var($search, FILTER_VALIDATE_EMAIL)) {
```
This should also not be done, when remote sharing has an exact match and vice-versa.
See old line 567 of the ShareesController code.

What would be the best to achive this? Or where is this done now

nickvergessen replied:
Ah it's in the Search class now, all good

nickvergessen commented on lib/private/Collaboration/Collaborators/SearchResult.php:
```diff
@@ -0,0 +1,97 @@
+			'groups' => [],
+			'remotes' => [],
+			'emails' => [],
+			'circles' => [],
```
I somewhat dislike this. Would be nicer if this is kept modular and provided by each app

blizzz replied:
In general, yes, I am not sure which side effects this might have for consumers of the endpoint.

OTOH it's better to risk a bug there now, than carrying cruft forever with us. I have a look at it again.

blizzz replied:
ah, i could simply merge it into the existing results array in the controller to preserver the state there, and have it leaner in the new impl.

nickvergessen commented:
Nice cleanup, I always disliked the naming anyway.
I would be fine to use a new route and just have two routes linking to the the same class/method.

blizzz commented:
> I would be fine to use a new route and just have two routes linking to the the same class/method.

From within cire, pointing to the files_sharing controller?  It might be disabled… We could move this, too, or think about a new Controller in the server… but this I see in a follow-up PR, rather.

blizzz committed:
don't keep result types hard coded

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
make it possible to register plugins	and kick out the circle one

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
allow more than one plugin per share type

however it does not dedupe (appears too complex/expensive while we don't
havve the issue currently)

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
finetune the appinfo parser

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
@nickvergessen  so I moved the Circles plugin to its own app (https://github.com/nextcloud/circles/pull/126) → registering Plugins works.

Also, the result types are not fixed but filled by the plugins. I added a public class for this to have type name validated. 

The LookupPlugin might go to the ~~Federation~~ files_sharing app (since it acts on its config), however this is a special case, because this is not activate by a provided share type, but a flag 😒 but this can be dealt with, with some extra effort.

Also, the MailPlugin can go to 'sharebymail' app, unless something speaks against it ( @schiessle ?), and the RemotePlugin to 'federation'.

nickvergessen commented on lib/public/Collaboration/Collaborators/SearchResultType.php:
```diff
@@ -0,0 +1,57 @@
+			throw new \InvalidArgumentException('Type must not be empty');
+		}
+
+		if(trim($type) === 'exact') {
```
trim is not necessary, you do it in the beginning already

blizzz replied:
was a leftover when i was pushing code around… thanks for pointing that out 👈 👍

blizzz committed:
updated autoloader

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
add missing PHP doc

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
minor fixes

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
split off getUsers unit tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
split off getGroups unit tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
split off getRemote unit tests

adds two small fixes → they actually work \o/

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
split off getEmail tests


Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
adjust ShareeAPIController search test

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
adjust ShareeAPIController invalid-input search unit test

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
split off former searchSharee unit test

also moves registering default plugins to Server for proper unit testing

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
adapt unit test for ensuring exception

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
move splitUserRemote unit tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
remove unused method

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
shareeAPIController unit test clean up and last fixes

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Now 🤹‍  with the ShareeAPIController unit tests was exciting 💤 ;)

blizzz commented on lib/private/Collaboration/Collaborators/MailPlugin.php:
```diff
@@ -159,6 +159,6 @@ public function search($search, $limit, $offset, ISearchResult $searchResult) {
 		$searchResult->addResultSet($emailType, $result['wide'], $result['exact']);
 
-		return false;
+		return true;
```
@schiessle this (= always more results available) feels wrong, but was required for the unit tests. Or did I oversee anything?

blizzz commented on lib/private/Collaboration/Collaborators/RemotePlugin.php:
```diff
@@ -116,7 +116,7 @@ public function search($search, $limit, $offset, ISearchResult $searchResult) {
 		$searchResult->addResultSet($resultType, $result['wide'], $result['exact']);
 
-		return false;
+		return true;
```
@schiessle this (= always more results available) feels wrong, but was required due to the unit tests. Or did I oversee anything?

blizzz committed:
add tests for Lookup Plugin

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
I kept LookupPlugin inside lib/…, because it is involved differently for whatever reason (invocation by explicit flag, not ShareType). Since there is no dedicated ShareType,  I refrain to change this.

This said, I'd be happy about final reviews 😸

daita commented:
Tested with Circles, working !

We need to edit _https://apps.nextcloud.com/schema/apps/info.xsd_ for
```
	<collaboration>
		<collaborators>
			<searchPlugins>
				<searchPlugin>
					<class></class>
					<shareType></shareType>
				</searchPlugin>
			</searchPlugins>
		</collaborators>
	</collaboration>
```

blizzz commented:
@daita yup, will do it next week

blizzz committed:
reduce nesting of the required app XML

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
due to requests from @nickvergessen and @BernhardPosselt i simplified the required XML. OK now? Needs green lights on https://github.com/nextcloud/appstore/pull/521, too.

blizzz committed:
adapt attribute name (changed letter case style)

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
🏓

nickvergessen commented:
Looks good and still works, but: current user is returned (was not the case before).
Maybe we can fix this before merging?

nickvergessen commented:
Okay lets do it in a follow up

nickvergessen commented on lib/private/Collaboration/Collaborators/MailPlugin.php:
```diff
@@ -0,0 +1,164 @@
+					$emailAddresses = [$emailAddresses];
+				}
+				foreach ($emailAddresses as $emailAddress) {
+					$exactEmailMatch = strtolower($emailAddress) === $lowerSearch;
```
For me this is an  empty string (when removing the email after it was set before), and therefor results in an entry `admin ()`
![bildschirmfoto vom 2017-10-04 15-24-54](https://user-images.githubusercontent.com/213943/31178083-424733de-a918-11e7-8365-421efd800954.png)

Of course this makes no sense, but we should have a look

blizzz replied:
I took it over as it is from the old method and have not spotted a post processing that would remove it. Thus, it should be reproducable with the old code I think… anyhow, yes, something we should fix
